src/dft/backends/cufft/backward.cpp

```diff
@@ -0,0 +1,179 @@
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+#include "oneapi/mkl/types.hpp"
+#include "oneapi/mkl/exceptions.hpp"
+
+#include "oneapi/mkl/dft/detail/cufft/onemkl_dft_cufft.hpp"
+#include "oneapi/mkl/dft/detail/types_impl.hpp"
+#include "oneapi/mkl/dft/detail/descriptor_impl.hpp"
+
+// cuFFT headers
+#include <cuda_runtime.h>
+#include <cufft.h>
+
+namespace oneapi::mkl::dft::cufft {
+namespace detail {
+
+template <dft::detail::precision prec, dft::detail::domain dom>
+inline dft::detail::commit_impl *get_commit(dft::detail::descriptor<prec, dom> &desc) {
+    auto commit_handle = dft::detail::get_commit(desc);
+    if (commit_handle == nullptr || commit_handle->get_backend() != backend::cufft) {
+        throw mkl::invalid_argument("DFT", "compute_backward",
+                                    "DFT descriptor has not been commited for cuFFT");
+    }
+    return commit_handle;
+}
+
+/// Throw an mkl::invalid_argument if the runtime param in the descriptor does not match
+/// the expected value.
+template <dft::detail::config_param Param, dft::detail::config_value Expected, typename DescT>
+inline auto expect_config(DescT &desc, const char *message) {
+    dft::detail::config_value actual{ 0 };
+    desc.get_value(Param, &actual);
+    if (actual != Expected) {
+        throw mkl::invalid_argument("DFT", "compute_backward", message);
+    }
+}
+} // namespace detail
+
+// BUFFER version
+
+//In-place transform
+template <typename descriptor_type, typename data_type>
+ONEMKL_EXPORT void compute_backward(descriptor_type &desc, sycl::buffer<data_type, 1> &inout) {
+    if constexpr (descriptor_type::precision == dft::detail::precision::SINGLE &&
+                  !std::is_floating_point_v<data_type>) {
+        detail::expect_config<dft::detail::config_param::PLACEMENT,
+                              dft::detail::config_value::INPLACE>(desc,
+                                                                  "Unexpected value for placement");
+        auto commit = get_commit(desc);
+        auto queue = commit->get_queue();
+        auto plan = *static_cast<cufftHandle *>(commit->get_handle());
+
+        cufftResult result = CUFFT_NOT_SUPPORTED;
+        cufftResult *result_ptr = &result;
+
+        queue.submit([&](sycl::handler &cgh) {
+            auto inout_acc = inout.template get_access<sycl::access::mode::read_write>(cgh);
+
+            cgh.host_task([=](sycl::interop_handle ih) {
+                auto inout_native = reinterpret_cast<cufftComplex *>(
+                    ih.get_native_mem<sycl::backend::ext_oneapi_cuda>(inout_acc));
+                auto stream = ih.get_native_queue<sycl::backend::ext_oneapi_cuda>();
+                *result_ptr = cufftSetStream(plan, stream);
+                if (*result_ptr == CUFFT_SUCCESS) {
+                    *result_ptr = cufftExecC2C(plan, inout_native, inout_native, CUFFT_INVERSE);
+                }
+            });
+        }).wait();
```

Comment:
This wait only waits on the host task to complete, BUT NOT on the device side work enqueued in the host task. 

From cuFFT documentation I do not find it clear if the execute calls are synchronous or asynchronous. The following paragraph assumes they are asynchronous.

Instead you should call cudaStreamSynchronize inside the host task. This will ensure the host task events and buffer synchronization actually match the device execution. 

If you move the following check on the result to the host task this wait becomes redundant.

Reply:
That was the intension since I only wanted to wait for result_ptr to be written to, not for the FFT to be executed. Anyway, I've thrown directly from the host task in https://github.com/oneapi-src/oneMKL/pull/278/commits/548a3d933e1501cbcac7135fec1beeb26724261f so that won't be an issue anymore.